includes/class-sensei-course-enrolment.php

```diff
@@ -0,0 +1,372 @@
+	/**
+	 * Check if a user is enrolled in a course.
+	 *
+	 * @param int  $user_id     User ID.
+	 * @param bool $check_cache Check and use cached result.
+	 *
+	 * @return bool
+	 */
+	public function is_enrolled( $user_id, $check_cache = true ) {
+		try {
+			if ( $check_cache ) {
+				$enrolment_check_results = $this->get_enrolment_check_results( $user_id );
+				if (
+					$enrolment_check_results
+					&& $enrolment_check_results->get_version_hash() === $this->get_course_enrolment_providers_version()
+				) {
+					return $this->has_stored_enrolment( $user_id );
+				}
+			}
+
+			$enrolment_check_results = $this->query_enrolment_providers( $user_id );
+			$is_enrolled             = $enrolment_check_results->is_enrolment_provided();
+
+			// @todo Method will always be bool when we add manual enrolment provider. Remove this and the `get_legacy_enrolment_status` temporary method.
+			if ( ! is_bool( $is_enrolled ) ) {
+				$is_enrolled = $this->get_legacy_enrolment_status( $user_id );
+			}
+
+			$this->save_enrolment( $user_id, $is_enrolled );
+		} catch ( Exception $e ) {
+			$is_enrolled = false;
+		}
+
+		return $is_enrolled;
+	}
+
+	/**
+	 * Trigger course enrolment check when enrolment might have changed.
+	 *
+	 * @param int $user_id User ID.
+	 */
+	public function trigger_course_enrolment_check( $user_id ) {
+		$this->is_enrolled( $user_id, false );
+	}
+
+	/**
+	 * Get enrolment from taxonomy record.
+	 *
+	 * @param int $user_id User ID.
+	 * @return bool
+	 * @throws Exception When learner term could not be created.
+	 */
+	private function has_stored_enrolment( $user_id ) {
+		$term = Sensei_Learner::get_learner_term( $user_id );
+
+		return has_term( $term->term_id, Sensei_PostTypes::LEARNER_TAXONOMY_NAME, $this->course_id );
+	}
+
+	/**
+	 * Save enrolment in taxonomy.
+	 *
+	 * @param int  $user_id    User ID.
+	 * @param bool $is_enrolled If the user is enrolled in the course.
+	 *
+	 * @return bool
+	 * @throws Exception When learner term could not be created.
+	 */
+	private function save_enrolment( $user_id, $is_enrolled ) {
+		$term = Sensei_Learner::get_learner_term( $user_id );
+		if ( ! $is_enrolled ) {
+			$result = wp_remove_object_terms( $this->course_id, [ intval( $term->term_id ) ], Sensei_PostTypes::LEARNER_TAXONOMY_NAME );
+
+			return true === $result;
+		}
+
+		// If they are enrolled, make sure they have started the course.
+		Sensei_Utils::start_user_on_course( $user_id, $this->course_id );
+
+		$result = wp_set_post_terms( $this->course_id, [ intval( $term->term_id ) ], Sensei_PostTypes::LEARNER_TAXONOMY_NAME, true );
+
+		return is_array( $result ) && ! empty( $result );
+	}
+
+	/**
+	 * Get the enrolment check results for a user.
+	 *
+	 * @access private Used internally only.
+	 *
+	 * @param int $user_id User ID.
+	 *
+	 * @return bool|Sensei_Course_Enrolment_Provider_Results
+	 * @throws Exception When learner term could not be created.
+	 */
+	public function get_enrolment_check_results( $user_id ) {
```

Comment:
Why make this `public`? I know we're marking it private, just curious since I'm not seeing it called anywhere outside this class.

Reply:
I realized when I started working with it that I wanted to inspect the results elsewhere, such as Learner Management.